modin/experimental/batch/test/test_pipeline.py

```diff
@@ -0,0 +1,750 @@
+            fan_out=True,
+            reduce_fn=reducer,
+            pass_partition_id=True,
+            is_output=True,
+        )
+        with pytest.raises(
+            NotImplementedError,
+            match="Fan out is only supported with DataFrames with 1 partition.",
+        ):
+            new_df = pipeline.compute_batch()[0]
+
+    def test_pipeline_complex(self):
+        import pandas
+        from os.path import exists
+        from os import remove
+
+        df = pd.DataFrame([[0, 1, 2]])
+
+        def new_col_adder(df, partition_id):
+            df["new_col"] = partition_id
+            return df
+
+        def reducer(dfs):
+            new_cols = "".join([str(df["new_col"].values[0]) for df in dfs])
+            dfs[0]["new_col1"] = new_cols
+            return dfs[0]
+
+        pipeline = PandasQueryPipeline(df, num_partitions=24)
+        pipeline.add_query(
+            new_col_adder,
+            fan_out=True,
+            reduce_fn=reducer,
+            pass_partition_id=True,
+            is_output=True,
+            output_id=20,
+        )
+        pipeline.add_query(
+            lambda df: pandas.concat([df] * 1000), repartition_after=True
+        )
+        pipeline.add_query(
+            lambda df: df.drop(columns=["new_col"]), is_output=True, output_id=21
+        )
+
+        def to_csv(df, partition_id):
+            df.to_csv(f"{partition_id}.csv")
+            return df
+
+        pipeline.add_query(to_csv, is_output=True, output_id=22, pass_partition_id=True)
+
+        def post_proc(df, o_id, partition_id):
+            df["new_col"] = f"{o_id} {partition_id}"
+            return df
+
+        new_dfs = pipeline.compute_batch(
+            postprocessor=post_proc,
+            pass_partition_id=True,
+            pass_output_id=True,
+            final_result_func=lambda df: df.iloc[-1],
+        )
+        correct_df = pd.DataFrame([[0, 1, 2]])
+        correct_df["new_col"] = "20 0"
+        correct_df["new_col1"] = "".join([str(i) for i in range(24)])
+        df_equals(correct_df.iloc[-1], new_dfs[20])
+        correct_df = pd.concat([correct_df] * 1000)
+        correct_df = correct_df.drop(columns=["new_col"])
+        correct_df["new_col"] = "21 0"
+        df_equals(correct_df.iloc[0], new_dfs[21])
+        for i in range(24):
+            assert exists(
+                f"{i}.csv"
+            ), "CSV File for Partition {i} does not exist, even though dataframe should have been repartitioned."
+            remove(f"{i}.csv")
+        assert 22 in new_dfs, "Output for output id 22 not generated."
+
+
+@pytest.mark.skipif(
+    Engine.get() == "Ray",
+    reason="Ray supports the Batch Pipeline API",
+)
+def test_pipeline_unsupported_engine():
+    """
+    This test ensures that trying to use the Pipeline API with an unsupported Engine raises errors.
+    """
+    # Check that pipeline does not allow `Engine` to not be Ray.
+    df = pd.DataFrame([[1, 2, 3], [4, 5, 6]])
```

Comment:
Let's keep it simple.

```suggestion
    df = pd.DataFrame([[1]])
```